Replace syn with handwritten TokenStream parser/output

ajtribick opened the pull request:
Since syn is somewhat infamous as a rather heavy crate in the build process, and the input/output of these macros are not too complex, here's a manual implementation.

- The implementation does NOT support escape sequences in string literals, which are of questionable utility here, but this is technically a breaking change.
- I've left in raw string literal support since this was included in the test cases already.

ajtribick committed:
Replace syn with handwritten TokenStream parser/output

- Reduces crate weight
- Breaking change: escape sequences are no longer parsed in literals

youknowone commented:
@lifthrasiir Could you review this PR? I feel like this is out of my review coverage

lifthrasiir commented on src/lib.rs:
```diff
@@ -44,13 +39,102 @@ pub fn hexf32(input: TokenStream) -> TokenStream {
+
+    let message_token = with_span(TokenTree::Literal(Literal::string(message)), span);
+
+    let mut tokens = TokenStream::new();
```
Please note the intended code as a comment.

```suggestion
    // ::core::compile_error!{<message_token>}
    let mut tokens = TokenStream::new();
```

lifthrasiir commented on src/lib.rs:
```diff
@@ -44,13 +39,102 @@ pub fn hexf32(input: TokenStream) -> TokenStream {
+    fn with_span(mut token_tree: TokenTree, span: Span) -> TokenTree {
+        token_tree.set_span(span);
+        token_tree
+    }
```
I think `span` is the only known span in this context, so a closure seems a better fit here.

lifthrasiir commented on src/lib.rs:
```diff
@@ -44,13 +39,102 @@ pub fn hexf32(input: TokenStream) -> TokenStream {
+fn extract_string_value(literal: &str) -> Option<&str> {
+    if literal.starts_with('"') {
+        if literal.len() >= 2 && literal.ends_with('"') {
+            // For simplicity, we do not handle any escape sequences. Given the
+            // usage of these macros, they would be of questionable utility.
```
I agree we shouldn't handle them here, but I think we should clearly indicate that escape sequences are explicitly not supported. I'm okay to reject the string if its code contains `\`. The same comment applies to the other branch.

ajtribick replied:
Added a specific error message when escape sequences are detected.

lifthrasiir commented on src/lib.rs:
```diff
@@ -44,13 +39,102 @@ pub fn hexf32(input: TokenStream) -> TokenStream {
+        Err(e) => compile_error_stream(&format!("hexf64! failed: {}", e), span),
+    })
+}
+
```
I would prefer to split the rest of this patch into a separate file, since a nitty-gritty of dealing with `TokenStream` shouldn't be the main concern here.

(I even feel that this parsing code has a bigger utility for other libraries and can be made into a small individual crate! If you feel like so and update this PR accordingly I will accept that too.)

ajtribick replied:
Moved to a separate file. I think there definitely is some utility to have a library that handles proc-macros which consist of a single string, on the other hand there is also some value in minimizing the dependencies of a crate like this. Might revisit this in the future if I end up putting in functionality like this in my own crate, for which a compile-time literal parser might also be useful (though I need to get round to writing proper parsing functionality at some point...).

lifthrasiir reviewed (changes requested):
I'm good with removing `syn` dependency. That said, I'm not very sure if this handles or at least recognizes some more legitimate edge cases like these:

```rust
hexf32!(    "0x0.0p0"     )
hexf32!(
    "0x0.00000000000000000000000000000000000000000000000000000000000p0"
) // can actually occur with rustfmt
```

I personally like to handle these two cases and reject any other edge cases, but I'll leave it up to you because these cases should be very rare anyway. The point is that if we don't support a particular syntax we should report why so that the user can rewrite it into the supported syntax.

ajtribick committed:
Code review feedback

- Move token stream handling into separate file
- Provide specific error message for escape sequences
- Add tests for surrounding whitespace

ajtribick commented:
Added some tests with whitespace, and addressed the other points.

lifthrasiir commented:
Thank you, and sorry for the late merger (I completely forgot to review the new commit for months until youknowone reminded me)!
